src/main/scala/mesosphere/marathon/core/task/Task.scala

```diff
@@ -83,39 +83,46 @@ object Task {
 
     private[this] def hasStartedRunning: Boolean = status.startedAt.isDefined
 
+    //scalastyle:off cyclomatic.complexity method.length
     override def update(update: TaskStateOp): TaskStateChange = update match {
       // case 1: now running
-      case TaskStateOp.MesosUpdate(MarathonTaskStatus.Running(mesosStatus), now) if !hasStartedRunning =>
+      case TaskStateOp.MesosUpdate(_, MarathonTaskStatus.Running(mesosStatus), now) if !hasStartedRunning =>
         val updated = copy(
           status = status.copy(
             startedAt = Some(now),
             mesosStatus = mesosStatus))
-        TaskStateChange.Update(updated)
+        TaskStateChange.Update(updated, Some(this))
 
       // case 2: terminal
-      case TaskStateOp.MesosUpdate(MarathonTaskStatus.Terminal(_), now) =>
-        TaskStateChange.Expunge
+      case TaskStateOp.MesosUpdate(_, MarathonTaskStatus.Terminal(_), now) =>
+        TaskStateChange.Expunge(this)
 
       // case 3: health or state updated
-      case TaskStateOp.MesosUpdate(taskStatus, now) =>
-        val healthOrStateChange = for {
-          current <- status.mesosStatus
-          update <- taskStatus.mesosStatus
-          newStatus <- updatedHealthOrState(current, update)
-        } yield TaskStateChange.Update(copy(status = status.copy(mesosStatus = Some(newStatus))))
-
-        healthOrStateChange.getOrElse{
+      case TaskStateOp.MesosUpdate(_, taskStatus, now) =>
+        updatedHealthOrState(status.mesosStatus, taskStatus.mesosStatus).map { newStatus =>
```

Comment:
Style (ignore me if you want)

if you handle both cases of an option, I prefer a `match`.